protocol/transaction_test.go

```diff
@@ -0,0 +1,43 @@
+package protocol
+
+import (
+	"bufio"
+	"bytes"
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestTransaction_Read(t *testing.T) {
+
+	buf := bytes.Buffer{}
+	comm := bufio.NewReadWriter(bufio.NewReader(&buf), bufio.NewWriter(&buf))
+	p := &Protocol{W: comm, R: comm, initialized: true}
+	trans := &transaction{p: p, in: []Message{}, out: []Message{}}
+
+	_, err := comm.Write([]byte{'P', 0, 0, 0, 4})
+	require.NoError(t, err)
+
+	err = comm.Flush()
+	require.NoError(t, err)
+
+	m, err := trans.Read()
+	require.NoError(t, err)
+	require.NotNilf(t, m,
+		"expected exactly 1 message in transaction incoming buffer. actual: %d", len(trans.in))
```

Comment:
This line is not exactly correct, the assertion does not look at `in` slice like the next one. Maybe it should be `expected a message from transaction, got nil` or similar.